### ISSUE TITLE
Analyzer input versioning

### ISSUE THREAD

**Issue by SleeplessByte** [exercism/automated-analysis#46]
Separated from https://github.com/exercism/automated-mentoring-support/issues/36#issuecomment-479845734

> As mentioned earlier on Slack, it would be helpful to get the package.json at the correct sha of the state of the exercise of the student, at time of the submission, because it will hold the current version. It's better than getting the sha directly, because we would need to keep track of an internal mapping (and therefore can not apply semver directly).

The _lowest_ we can support is getting the input exercise `git sha` and manually build a version mapping. This can be provided:
- as an input argument (`--sha="abcdef"`)
- as a file (`.git/version` or similar)

Better ways:
- provide the track specific version file (e.g. `version`, `package.json`, or similar) along the submitted solution
- provide *all* the files as if you did `exercism download` for the specific solution (so that we can fetch the version ourselves)

We need this information in order to be able to analyze _based on the version of the input_ instead of `LATEST`. This makes it very straightforward to actively support or **not support** legacy exercises.

**Comment by tehsphinx** [comment]
> as an input argument (--sha="abcdef")

I like this one most as it keeps the effort and the dependencies in all the analyzers small.

If no version is provided, the newest should be assumed.

**Comment by SleeplessByte** [comment]
> If no version is provided, the newest should be assumed.

Yes!

**Comment by iHiD** [comment]
If we go down the git-sha root, we could also potentially use the most recent git-sha when an **exercise** directory was touched, lower than or equal to the git-sha that the solution is locked at. git-shas then effectively become mapped 1-1 to the exercises.

Does that make any sense?

**Comment by SleeplessByte** [comment]
Yes, but how would I know what the _order_ of git-sha's is?

**Comment by iHiD** [comment]
So, let's say the repo has 20 commits, handily numbered 1-20
- Exercise 1 was last updated in commits [7, 9 and 13].
- User 1's solution is locked to git-sha-12, so I would pass git-sha-9 to you
- User 2's solution is up to date (locked to git-sha-20), so I would pass git-sha-13 to you

So rather than passing the solution sha to you, I pass the last exercise sha to you.

The analyzer can then say "I know that exericse sha - here's what I do with that" or it can say "Hmmm, I don't know that sha, I'll just go with latest".

**Comment by NobbZ** [comment]
What do you consider updates of the exercises here?

Similar to how the website currently does consider it an update when files have been touched that match the test-pattern, or when any file changed in the exercise?

**Comment by iHiD** [comment]
I think any file changing within the exercise directory.

**Comment by iHiD** [comment]
> provide all the files as if you did exercism download for the specific solution (so that we can fetch the version ourselves)

We can do this and/or the config.json file. Let's open a new issue for this once the new pass of analyzers is done.
